zigzag/zigzag.py

```diff
@@ -13,259 +13,274 @@
-        junit_xml = junit_xml_doc.getroot()
-    except etree.DocumentInvalid as e:
-        debug = "\n\n---DEBUG XML PRETTY PRINT---\n\n"
-        error_message = "The file '{}' does not conform to schema!" \
-                        "\n\nSchema Violation:\n{}".format(file_path, str(e))
-        if pprint_on_fail:
-            error_message = "{0}{1}{2}{1}".format(error_message,
-                                                  debug,
-                                                  etree.tostring(junit_xml_doc, pretty_print=True))
-        raise RuntimeError("The file '{}' does not conform to schema!"
-                           "\n\nSchema Violation:\n{}".format(file_path, error_message))
+        Raises:
+            KeyError: missing test suite property.
+            AttributeError: the testcase 'classname' attribute is invalid
+        """
 
-    if junit_xml.tag != root_element:
-        raise RuntimeError("The file '{}' does not have JUnitXML '{}' root element!".format(file_path, root_element))
+        module_hierarchy = [testsuite_props['RPC_RELEASE'],             # RPC Release Version (e.g. 16.0.0)
+                            testsuite_props['JOB_NAME'],                # CI Job name (e.g. PM_rpc-openstack-pike-xenial_mnaio_no_artifacts-swift-system) # noqa
+                            testsuite_props['MOLECULE_TEST_REPO'],      # (e.g. molecule-validate-neutron-deploy)
+                            testsuite_props['MOLECULE_SCENARIO_NAME']]  # (e.g. "default")
 
-    return junit_xml
+        testcase_groups = self._TESTCASE_GROUP_RGX.search(testcase_xml.attrib['classname']).groups()
 
+        module_hierarchy.append(testcase_groups[0])         # Always append at least the filename of the test grouping.
+        if testcase_groups[1]:
+            module_hierarchy.append(testcase_groups[1])     # Append the class name of tests if specified.
 
-def _generate_module_hierarchy(testcase_xml, testsuite_props):
-    """Construct a qTest swagger model for all the JUnitXML test cases.
+        return module_hierarchy
 
-    Args:
-        testcase_xml (ElementTree): A XML element representing a JUnit style testcase result.
-        testsuite_props (dict): Global properties from the associated testsuite for the given testcase result.
+    def _generate_test_logs(self):
+        """Construct a qTest swagger model for all the JUnitXML test cases.
 
-    Returns:
-        list(str): An ordered list of strings to use for the qTest results hierarchy.
+        Returns:
+            list(AutomationTestLogResource): A list of qTest swagger model test logs.
+        """
 
-    Raises:
-        KeyError: missing test suite property.
-        AttributeError: the testcase 'classname' attribute is invalid
-    """
+        serialized_junit_xml = etree.tostring(self._junit_xml, encoding='UTF-8', xml_declaration=True)
+        testsuite_props = {p.attrib['name']: p.attrib['value'] for p in self._junit_xml.findall('./properties/property')}  # noqa
```

Comment:
You can avoid the `# noqa` mark by placing a line-break right before the `for` statement.

```python
testsuite_props = {p.attrib['name']: p.attrib['value']
                             for p in self._junit_xml.findall('./properties/property')}
```

Reply:
fixed